### ISSUE TITLE
Add Test New Script functionality to tests

### ISSUE THREAD

**Issue by ethanwhite** [weecology/retriever#741]
As we scale up the number of datasets a common issue will be testing new scripts both locally and through CI. A nice feature would be a test that detects new scripts (by comparing the local script list against those listed in `version.txt` on `master`) and performs an equivalent to the `try_run_all.py` code for those scripts.

For small to medium datasets this should be runable on CI. For large datasets we might need to add a `donottest` or `slow` flag that allows control over whether they are run or not (using pytest's `--runslow` option).

**Comment by ShivamNegi** [comment]
What about the scripts which are behind the version mentioned in the `version.txt`. There are three instances of these namely `Car.json`, `WineQ.json`, `BreastCancer.json`. Could you explain the part of what is to be done with the `try_run_all.py`.

**Comment by ethanwhite** [comment]
> What about the scripts which are behind the version mentioned in the version.txt.

That shouldn't generally happen in the repository. Because when a script is updated the `version` should be incremented and `version.py` should be rerun. I can see accidents happening in the presence of merge conflicts, but in this case...

> There are three instances of these namely Car.json, WineQ.json, BreastCancer.json.

Following da37cd these files are no longer in `master`. If you're still seeing them I would recommend doing the following if you have this repository set up as `upstream` and don't have any uncommitted changes:

```
git fetch upstream
git merge upstream master
rm -rf scripts
git reset --hard HEAD
```

This should get you back in line with `master`. This was a big renaming effort and it shouldn't happen again.

> Could you explain the part of what is to be done with the try_run_all.py.

`try_run_all.py` is designed to allow us to attempt to install all of the datasets into all of the different output formats (called "engines"). It then returns a list of any errors that occurred so that we can debug them.

The goal of this issue is to take any scripts that have been updated and try to install them into all the output formats to make sure that the new script is working, or at least not throwing any errors.

**Comment by ShivamNegi** [comment]
I am having a problem with the script file names. While making the test script for checking script versions, I am importing the scripts under the `retriever/scripts` folder. Due the the file names having a `-` character in them python is not able to import these file. eg. `breed-bird-survey.py`. I have found some workarounds for this but I think it would be better to rename all the files with proper naming.

**Comment by ethanwhite** [comment]
Can you provide a specific example of the problems you're running into? Python doesn't allow `-` in variable names, but it should be fine with reading files with those names. I'm open to using `_` if it's a real problem, but `-` is more common in the package managers I've worked with.

**Comment by ShivamNegi** [comment]
I was thinking of importing the scripts for testing. If reading is a better way to get the script versions then there is no problem.

**Comment by ethanwhite** [comment]
Ah, I see your point now. Thanks. We are importing these, but we must be doing it in a non-standard way since it's handling the dashes.

@henrykironde - what are your thoughts on this. Since these are .py files should we rename them with `_` so they can be imported in the usual way?

**Comment by henrykironde** [comment]
I am not really sure the exact problem @ShivamNegi is facing. Could you explain to me the issue in detail. For the naming convention we could follow the recommended use of `_` rather than `-` . And if that solves the problem you are facing, go ahead and make a PR.

**Comment by ShivamNegi** [comment]
`import breed-bird-survey`
gives an error in python due to the presence of `-` in the filename. Though the script can be import using `__import__`, which is non-standard.

**Comment by henrykironde** [comment]
Okey I get it thanks

**Comment by ethanwhite** [comment]
OK, so the recommendation for [naming Python modules from PEP 8](https://www.python.org/dev/peps/pep-0008/#package-and-module-names) is:

> Modules should have short, all-lowercase names. Underscores can be used in the module name if it improves readability. 

So, my suggestion at this point is to use `-` in the `name` field in the module (since that's standard for naming in package management systems) and use `_` in place of `-` for the filenames to follow Python convention.

Thoughts?

**Comment by ShivamNegi** [comment]
I agree to naming with `_`. PEP 8 naming conventions also say so.
https://www.python.org/dev/peps/pep-0008/#package-and-module-names

**Comment by ShivamNegi** [comment]
I have only made the test script till now. 
> A nice feature would be a test that detects new scripts (by comparing the local script list against those listed in version.txt on master) and performs an equivalent to the try_run_all.py code for those scripts.

@ethanwhite where you referring to `try_install_all.py`?
The function that would be equivalent to the `try_install_all.py` should be made as a new script? Or should I add it in the test file itself.

**Comment by henrykironde** [comment]
` try_install_all.py` , gets all the scripts and runs them over the data base engines. 
We shall take the design from ` try_install_all.py` and provide the new function/test with the value of the new scripts.
In other words, the new test module/function you are designing should have most of the part as in ` try_install_all.py`, but it will instead take a list of values(these are the new scripts) as arguments

1. should be able to detect the new scripts.

- In this case it could use the `version.txt` to detect the new scripts, So if there is a new script add that to the list. The function will run test using the various engines on every new script.

 
2. [try_install_all.py](https://github.com/weecology/retriever/blob/master/try_install_all.py) , we shall have to change a few things and  the starting point would be,
Change the `IGNORE` to the list of modified scripts. i.e `modified_scripts`
Then
`if module.SCRIPT.shortname.lower() not in IGNORE:` to 
`if module.SCRIPT.shortname.lower() in modified_scripts:`

With the changes you will be left to clean up the code and then create a test_function as appropriate
```
SCRIPT_LIST = SCRIPT_LIST()
TEST_ENGINES = {}
**modified_scripts** = [ "NEW SCRIPTS" ]

for engine in ENGINE_LIST:
    if engine.abbreviation in engine_test:
        try:
            opts = engine_test[engine.abbreviation]
            TEST_ENGINES[engine.abbreviation] = choose_engine(opts)
        except:
            TEST_ENGINES[engine.abbreviation] = None
            pass

errors = []
for module in MODULE_LIST:
    for (key, value) in list(TEST_ENGINES.items()):
        **if module.SCRIPT.shortname.lower() not in IGNORE:**
            if value != None:
                print("==>", module.__name__, value.name, "..........", module.SCRIPT.shortname)
                try:
                    module.SCRIPT.download(value)
                except KeyboardInterrupt:
                    pass
                except Exception as e:
                    print("ERROR.")
                    errors.append((key, module.__name__, e))
            else:
                errors.append((key, "No connection detected......" + module.SCRIPT.shortname))
```

**Comment by ShivamNegi** [comment]
Thanks. I see that the `version.txt` has been changed. Which now does not conform to the `version.py` output. Can you tell me from where this new `version.txt` is being created?

**Comment by ethanwhite** [comment]
@ShivamNegi - I suspect that what's happening is that you have some old scripts cached that have now been renamed. Can you run `retriever reset scripts` and then try running `version.py` again and see if that cleans things up?

**Comment by ethanwhite** [comment]
Oh, and you'll need to run `retriever update` as well. So:

```
retriever reset scripts
retriever update
python version.py
```

**Comment by ShivamNegi** [comment]
I am unsure on how to structure the script. Should it be divided into two separate scripts. One under /test as the test script which checks script versions and other as `./try_install_modified.py`. Where the `try_install_modified.py` is called when test fails for modified scripts.

**Comment by ethanwhite** [comment]
Implemented in 2293c2ee7194c767624a4ab1134b8283a2c59b2b. Thanks again for this @ShivamNegi!
